File: java/omnia/data/cache/SimpleDoubleCacher.java
package omnia.data.cache;

import static java.util.Objects.requireNonNull;

import java.util.function.DoubleSupplier;

/**
 * A {@link CachedDouble} implementation that uses the given {@link DoubleSupplier} to provide the
 * value to be cached. This class is not thread-safe.
 *
 * The given {@link DoubleSupplier#getAsDouble()} method is never invoked if the value is cached.
 */
final class SimpleDoubleCacher implements CachedDouble {
  private final DoubleSupplier supplier;
  private boolean isValid;
  private double value;

  SimpleDoubleCacher(DoubleSupplier supplier) {
    this.supplier = requireNonNull(supplier);
  }

  @Override
  public double value() {
    if (!isValid) {
      value = supplier.getAsDouble();
      isValid = true;
    }
    return value;
  }

  @Override
  public void invalidate() {
    isValid = false;
  }
}

File: src/omnia/data/cache/SimpleDoubleCacher.kt
package omnia.data.cache

import java.util.function.DoubleSupplier

/**
 * A [CachedDouble] implementation that uses the given [DoubleSupplier] to provide the
 * value to be cached. This class is not thread-safe.
 *
 * The given [DoubleSupplier.getAsDouble] method is never invoked if the value is cached.
 */
internal class SimpleDoubleCacher(private val supplier: DoubleSupplier) : CachedDouble {

  private var isValid = false
  private var value = 0.0
  override fun value(): Double {
    val value = if (isValid) this.value else supplier.asDouble
    isValid = true
    this.value = value
    return value
  }

  override fun invalidate() {
    isValid = false
  }
}